### ISSUE TITLE
Storing your posts on Swarm

### ISSUE THREAD

**Issue by agazso** [agazso/postmodern#144]
In this issue I would like to discuss the different options about storing all your shared posts. When thinking about this problem I came up with two criteria that a good solution would satisfy:

1. There is an ordering of the posts and if you a link to a newer post, you can access all the previous posts from it, so you can build a timeline of all posts.
2. It's possible to remove or update a post from the timeline without having to update all the newer posts.

The goal of the 1st one is discoverability, that is when you access a post from someone you can also see their other posts, follow the links from them etc.
The goal of the 2nd one is to avoid having to generate traffic (that costs money) when you are trying to remove or update older posts.

As a conversation starter I have two proposals which satisfies both criteria.

### Proposal 1: Storing the posts Swarm manifests

The idea is that we store the posts in a Swarm _manifest_ (or more precisely the hashes of the posts, because that's how things are stored in a manifest). Then we store the hash of the manifest in a Swarm Feed (`AllPostFeed`). Every time the manifest changes, we update its hash in the `AllPostFeed`.

With this we solved the 2nd goal because we can indeed remove or update posts, all we have to do is to update the manifest and then update the `AllPostFeed` too. For solving the 1st goal you need to maintain an ordering of the posts separately from the posts. For that we can use the following model:

In each post we store a link to the `AllPostFeed`. With that it will be possible to access all posts. Here is how: for each post (identified by its hash) we maintain a mapping between a post and its previous post. So in the manifest we create entries that are named by a hash of the post and contains the hash of the previous post and put the under a certain path on Swarm. We also maintain an entry for the newest post. Then there are two cases:

In both cases we start from this state:

`newestPost` -> `hash-of-postC`
`hash-of-PostC` -> `hash-of-PostB`
`hash-of-PostB` -> `hash-of-PostA`

#### Case 1: Updating an item

Let's say we want to update `PostB` therefore we create `PostBUpdate`. We want to modify our structure so that it is updated accordingly:

`newestPost` -> `hash-of-postC`
**`hash-of-PostC` -> `hash-of-PostBUpdate`**
**`hash-of-PostBUpdate` -> `hash-of-PostA`**
_`hash-of-PostB` -> `hash-of-PostA`_

The lines marked with **bold** are the new updates and the line marked with _italic_ is no longer needed so it can be removed safely. With that updating one post requires adding only two items to the manifests (and optionally removing one), so it's a fixed cost that does not depend on the location, therefore it does not violate our 2nd goal.

#### Case 2: Removing an item

Let's say we want to remove `PostB` from the timeline. The updated structure looks like this.

`newestPost` -> `hash-of-postC`
**`hash-of-PostC` -> `hash-of-PostA`**
_`hash-of-PostB` -> `hash-of-PostA`_

The lines marked with **bold** are the new updates and the line marked with _italic_ is no longer needed so it can be removed safely. With that removing one post requires having one update to the manifests (and optionally removing one), so it's a fixed cost that does not depend on the location, therefore it does not violate our 2nd goal.

#### Conclusion of using Swarm manifests

As demonstrated it satisfies our two initial criteria and also a bit more, because it enables to access all the posts from a certain post, not just the older ones. However there might be other aspects that we may have to take into consideration:

- Cost/performance: we don't yet know what's the cost (in time and money) to work with manifests. For example in order to query the full timeline and if one has to query each item sequentially and if the roundtrip is big (for example when using a gateway) it can add up easily. Think of 600ms RTT and just fetching the hashes of the last 20 posts would take 12 seconds, and we haven't even downloaded the posts (although that can be started in parallel). We need to do further research on how `bzz-list` works.

- There are possible consistency issues, because the data and the mapping is separated and because there are no transactions or atomic update in Swarm. This can cause problems when the connectivity is lost during an update or when multiple users updating the manifest at the same time. This is a general problem with Swarm and it can be mitigated by executing the update in the correct ordering so at least there are no dangling 'pointers' left breaking the timeline.

### Proposal 2: Using Swarm Feeds for posts

In this proposal each post have its own Swarm Feed (`PostChangeFeed`). In each post, we store a link to the previous post's `PostChangeFeed` and we also store a link to a Swarm Feed (`AllPostFeed`) that stores the link to the newest post (This is very similar this way to a single-linked list where each item have a pointer to the actual head).

So this satisfies our 1st goal, because from each post we can access all previous post and also the new ones too. How do we update and remove items?

Let's say we start from this state:

`AllPostFeed` -> `PostChangeFeedC`
`PostChangeFeedC` -> `PostC` -> `PostChangeFeedB`
`PostChangeFeedB` -> `PostB` -> `PostChangeFeedA`
`PostChangeFeedA` -> `PostA`

#### Case 1: Updating an item

Let's say we want to update `PostB` therefore we create `PostBUpdate`. We want to modify our structure so that it is updated accordingly:

`AllPostFeed` -> `PostChangeFeedC`
`PostChangeFeedC` -> `PostC` -> `PostChangeFeedB`
**`PostChangeFeedB` -> `PostBUpdate` -> `PostChangeFeedA`**
`PostChangeFeedA` -> `PostA`

We only had to modify `PostChangeFeedB` so that it points to the updated version and make sure that in the updated version we link to the previously linked change feed (`PostChangeFeedA`).

#### Case 2: Removing an item

Let's say we want to remove `PostB` from the timeline. The updated structure looks like this.

`AllPostFeed` -> `PostChangeFeedC`
**`PostChangeFeedC` -> `PostCUpdate` -> `PostChangeFeedA`**
`PostChangeFeedA` -> `PostA`
_`PostChangeFeedB` -> `PostB` -> `PostChangeFeedA`_

We had to update `PostChangeFeedC` and `PostC` and then remove (or ignore) `PostChangeFeedB`.

#### Conclusion of using Swarm Feeds

It seems that using Swarm Feeds are conceptually simpler, because in both cases you only have to update one post and one Swarm Feed. However it may be much slower to access the full timeline, because you always have to lookup the feed, then find the current version (which may be also several lookups) and then fetch the post to get the address of the feed of the previous post and continue.

**Comment by vujevits** [comment]
### Proposal 3: Storing operations in Swarm feed

In this case we would define 3 operations
- add
- update
- remove

I might be wrong, but this solution might be more flexible regarding future changes. We can introduce special operations e.g. for migration, because the resolution of update/removals is already done on the client side.

Starting from here:
`Post3(add)`
`Post2(add)`
`Post1(add)`

#### Case 1: updating item
**`Post2(update)`**
`Post3(add)`
`Post2(add)`
`Post1(add)`


#### Case 2: removing item
**`Post2(remove)`**
`Post3(add)`
`Post2(add)`
`Post1(add)`

**Comment by agazso** [comment]
I am not sure I completely understand your proposal. What I understand it's kind of an Event Sourcing model, which is indeed good if you want your updates to be quick, however it has a bigger cost when reading or querying the data. Also it might need some kind of Garbage Collection or compaction mechanism to avoid

**Comment by agazso** [comment]
What do you mean here by feed? 

I am talking about how to find a way that storing the posts that can be implemented by the Swarm APIs, and in this case Swarm Feeds. In a Swarm Feed you can store any kind of byte array as data, and then later you can update this data, but it's address will stay the same.

This means that Swarm has no idea what data you store, therefore if you want to impose any kind of constraints and operations on your data, you have to define it first. That's what my question is referring that in your case what would be the information that we store in the Swarm Feed.

**Comment by vujevits** [comment]
By feed I mean Swarm Feed. In case we store the necessarry information locally as well, you don't need the whole history to download, just until the point of the last 'sync'
